libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+			r_debug_lstLibAdd (pid,de.u.LoadDll.lpBaseOfDll, de.u.LoadDll.hFile, dllname);
+			if (dllname) {
+				free (dllname);
+			}
+			next_event = 0;
+			ret = R_DEBUG_REASON_NEW_LIB;
+			break;
+		case UNLOAD_DLL_DEBUG_EVENT:
+			//eprintf ("(%d) Unloading library at %p\n", pid, de.u.UnloadDll.lpBaseOfDll);
+			lstLibPtr = (PLIB_ITEM)r_debug_findlib (de.u.UnloadDll.lpBaseOfDll);
+			if (lstLibPtr != NULL) {
+				lstLibPtr->hFile = INVALID_HANDLE_VALUE;
+			} else {
+				r_debug_lstLibAdd (pid, de.u.UnloadDll.lpBaseOfDll, INVALID_HANDLE_VALUE, "not cached");
+				if (dllname)
+					free (dllname);
+			}
+			next_event = 0;
+			ret = R_DEBUG_REASON_EXIT_LIB;
+			break;
+		case OUTPUT_DEBUG_STRING_EVENT:
+			//eprintf ("(%d) Debug string\n", pid);
+			r_cons_printf ("(%d) Debug string\n", pid);
+			r_cons_flush ();
+
+			w32_continue (dbg, pid, tid, -1);
+			next_event = 1;
+			break;
+		case RIP_EVENT:
+			//eprintf ("(%d) RIP event\n", pid);
+			r_cons_printf ("(%d) RIP event\n", pid);
+			r_cons_flush ();
+			w32_continue (dbg, pid, tid, -1);
+			next_event = 1;
+			// XXX unknown ret = R_DEBUG_REASON_TRAP;
+			break;
+		case EXCEPTION_DEBUG_EVENT:
+			switch (de.u.Exception.ExceptionRecord.ExceptionCode) {
+#if _WIN64
+			case 0x4000001f: /* STATUS_WX86_BREAKPOINT */
+#endif
+			case EXCEPTION_BREAKPOINT:
+				ret = R_DEBUG_REASON_BREAKPOINT;
+				next_event = 0;
+				break;
+#if _WIN64
+			case 0x4000001e: /* STATUS_WX86_SINGLE_STEP */
+#endif
+			case EXCEPTION_SINGLE_STEP:
+				ret = R_DEBUG_REASON_STEP;
+				next_event = 0;
+				break;
+			default:
+				if (!debug_exception_event (&de)) {
+					ret = R_DEBUG_REASON_TRAP;
+					next_event = 0;
+				}
+				else {
+					next_event = 1;
+					w32_continue (dbg, pid, tid, -1);
+				}
+
+			}
+			break;
+		default:
+			eprintf ("(%d) unknown event: %d\n", pid, code);
+			return -1;
+		}
+	} while (next_event);
+	return ret;
+}
+
+int w32_step(RDebug *dbg) {
+	/* set TRAP flag */
+	CONTEXT ctx;
+	w32_reg_read (dbg, R_REG_TYPE_GPR, (ut8 *)&ctx, sizeof (ctx));
+	ctx.EFlags |= 0x100;
+	w32_reg_write (dbg, R_REG_TYPE_GPR, (ut8 *)&ctx, sizeof (ctx));
+	w32_continue (dbg, dbg->pid, dbg->tid, dbg->reason.signum);
+	// (void)r_debug_handle_signals (dbg);
+	return true;
+}
+
+int w32_continue(RDebug *dbg, int pid, int tid, int sig) {
+	/* Honor the Windows-specific signal that instructs threads to process exceptions */
+	DWORD continue_status = (sig == DBG_EXCEPTION_NOT_HANDLED)
+		? DBG_EXCEPTION_NOT_HANDLED : DBG_CONTINUE;
+	if (ContinueDebugEvent (pid, tid, continue_status) == 0) {
+		r_sys_perror ("w32_continue/ContinueDebugEvent");
+		eprintf ("debug_contp: error\n");
+		return false;
+	}
+	return tid;
+}
+
+RDebugMap *w32_map_alloc(RDebug *dbg, ut64 addr, int size) {
+	RIOW32Dbg *rio = dbg->user;
+	LPVOID base = VirtualAllocEx (rio->ph, (LPVOID)(size_t)addr,
+	  			(SIZE_T)size, MEM_COMMIT, PAGE_READWRITE);
+	if (!base) {
```

Comment:
i would be positive here and do `if (base) { ..` return NULL;

Reply:
I'm not entirely sure I understand what you mean by saying this.